lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -772,6 +772,63 @@ def accessible_objects(user_or_attribute, operation, options = {})
       end
     end
 
+    # Version of accessible_objects which only returns objects that are
+    # ancestors of a specified root object or the object itself
+    def accessible_ancestor_objects(user_or_attribute, operation, root_object, options = {})
+      # If the root_object is a generic PM::Object, convert it the appropriate storage adapter Object
+      root_object_stored_pe = root_object.try(:stored_pe) || root_object
+
+      # The final set of accessible objects must be ancestors of the root_object
+      ancestor_objects = root_object_stored_pe.ancestors(type: class_for_type('object')) + [root_object_stored_pe]
+
+      # Short-circuit and return all ancestors (minus prohibitions) if the user_or_attribute is authorized on the root node
+      if is_privilege?(user_or_attribute, operation, root_object_stored_pe)
+        apply_include_condition!(ancestor_objects, options[:key], options[:includes])
+
+        if !options[:ignore_prohibitions] && prohibition = prohibition_for(operation)
+          prohibited_ancestor_objects = accessible_ancestor_objects(
+            user_or_attribute,
+            prohibition,
+            root_object_stored_pe,
+            options.merge(ignore_prohibitions: true)
+          )
+          return ancestor_objects - prohibited_ancestor_objects
+        else
+          return ancestor_objects
+        end
+      end
+
+      # Fetch all of the PEAs using the given UA or its descendants
+      user_attribute_ids = user_or_attribute.descendants.pluck(:id) | [user_or_attribute.id]
+      associations = PolicyElementAssociation.where(user_attribute_id: user_attribute_ids)
+      operation_set_ids = associations.pluck(:operation_set_id)
+
+      # Narrow the list of PEAs to just those containing the specified operation
+      operation_id = operation.try(:unique_identifier) || operation.to_s
+      filtered_operation_set_ids = Assignment.filter_operation_set_list_by_assigned_operation(operation_set_ids, operation_id)
+      filtered_associations = associations.where(operation_set_id: filtered_operation_set_ids)
+
+      permitting_oa_ids = filtered_associations.map(&:object_attribute_id)
```

Comment:
Why do a map here but a pluck above? Does pluck construct its own sql query above? If so, we should probably use map in both places.

Reply:
I'll `pluck` in both.